File: C++/2-Medium/00310._最小高度树/00310._最小高度树.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <list>
using namespace std;

class Solution {
public:
    vector<int> findMinHeightTrees(int n, vector<vector<int>>& edges) {
        if (n == 1) return {0}; // 边界情况：只有一个节点
        vector<list<int>> adj(n); // 邻接表表示图
        vector<int> degree(n, 0); // 节点的度
        for (auto& e : edges) { // 建图
            adj[e[0]].push_back(e[1]);
            adj[e[1]].push_back(e[0]);
            degree[e[0]]++;
            degree[e[1]]++;
        }
        
        queue<int> q; // 存储度数为1的节点（叶子节点）
        for (int i = 0; i < n; ++i) {
            if (degree[i] == 1) q.push(i);
        }
        
        while (n > 2) {
            int size = q.size();
            n -= size; // 删除叶子节点后，更新图中节点的数量
            for (int i = 0; i < size; ++i) {
                int t = q.front(); q.pop();
                for (auto it = adj[t].begin(); it != adj[t].end(); ++it) {
                    if (--degree[*it] == 1) q.push(*it); // 更新度数，并检查是否变成叶子节点
                }
            }
        }
        
        vector<int> result;
        while (!q.empty()) {
            result.push_back(q.front());
            q.pop();
        }
        return result;
    }
};

int main() {
    Solution solution;
    int n = 4;
    vector<vector<int>> edges = {{1, 0}, {1, 2}, {1, 3}};
    vector<int> result = solution.findMinHeightTrees(n, edges);
    for (int root : result) {
        cout << root << " ";
    }
    cout << endl;
    return 0;
}

File: Python/2-Medium/00310._最小高度树/00310._最小高度树.py
from typing import List, Dict, Set

def findMinHeightTrees(n: int, edges: List[List[int]]) -> List[int]:
    """
    找到所有最小高度树的根节点。

    Args:
    n (int): 节点的数量。
    edges (List[List[int]]): 边的列表，每个元素是一对节点。

    Returns:
    List[int]: 所有最小高度树的根节点列表。
    """
    if n == 1:  # 如果只有一个节点，直接返回这个节点
        return [0]
    # 构建图的邻接表和每个节点的度
    graph = [set() for _ in range(n)]
    degree = [0] * n
    for u, v in edges:
        graph[u].add(v)
        graph[v].add(u)
        degree[u] += 1
        degree[v] += 1
    
    # 找到所有初始的叶子节点
    leaves = [i for i in range(n) if degree[i] == 1]
    
    # 逐层剥离叶子节点，直到剩下的节点数小于等于2
    while n > 2:
        n -= len(leaves)  # 更新剩余节点的数量
        new_leaves = []
        for leaf in leaves:
            neighbor = graph[leaf].pop()  # 获取叶子节点的邻居并从图中删除叶子节点
            graph[neighbor].remove(leaf)
            degree[neighbor] -= 1  # 更新邻居的度
            if degree[neighbor] == 1:  # 如果邻居变成叶子节点，加入到新的叶子节点列表中
                new_leaves.append(neighbor)
        leaves = new_leaves  # 准备下一轮的叶子节点剥离
    
    return leaves  # 返回剩下的节点作为最小高度树的根节点

# 测试用例
n = 4
edges = [[1, 0], [1, 2], [1, 3]]
# 运行代码并打印结果
print(findMinHeightTrees(n, edges))